ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,490 @@
+    cdef ucp_config_t *config = _read_ucx_config({})
+    ret = ucx_config_to_dict(config)
+    ucp_config_release(config)
+    return ret
+
+
+def get_ucx_version():
+    cdef unsigned int a, b, c
+    ucp_get_version(&a, &b, &c)
+    return (a, b, c)
+
+
+cdef void _listener_callback(ucp_ep_h ep, void *args) except *:
+    cdef dict cb_data = <dict> args
+    cb_data['cb_func'](
+        ucx_ep_create(ep),
+        *cb_data['cb_args']
+    )
+
+
+cdef class UCXListener:
+    cdef:
+        ucp_listener_h handle
+        uint16_t _port
+        dict cb_data
+        bint initialized
+
+    def __cinit__(
+        self,
+        UCXWorker worker,
+        uint16_t port,
+        cb_func,
+        cb_args
+    ):
+
+        self.initialized = False
+        self._port = port
+        self.cb_data = {
+            'cb_func': cb_func,
+            'cb_args': cb_args,
+        }
+
+        cdef ucp_listener_params_t params
+        cdef ucp_listener_accept_callback_t _listener_cb = (
+            <ucp_listener_accept_callback_t>_listener_callback
+        )
+        if c_util_get_ucp_listener_params(&params,
+                                          port,
+                                          _listener_cb,
+                                          <void*> self.cb_data):
+            raise MemoryError("Failed allocation of ucp_ep_params_t")
+
+        logging.info("create_listener() - Start listening on port %d" % port)
+        cdef ucs_status_t status = ucp_listener_create(
+            worker._handle, &params, &self.handle
+        )
+        c_util_get_ucp_listener_params_free(&params)
+        assert_ucs_status(status)
+        self.initialized = True
+
+    def port(self):
+        return self._port
+
+    def destroy(self):
+        if self.initialized:
+            self.initialized = False
+            ucp_listener_destroy(self.handle)
+
+
+cdef class UCXContext:
+    cdef:
+        ucp_context_h _context
+        bint _initialized
+        object _config
+
+    def __cinit__(self, config_dict):
+        cdef ucp_params_t ucp_params
+        cdef ucp_worker_params_t worker_params
+        cdef ucs_status_t status
+        self._initialized = False
+
+        memset(&ucp_params, 0, sizeof(ucp_params))
+        ucp_params.field_mask = (UCP_PARAM_FIELD_FEATURES |  # noqa
+                                 UCP_PARAM_FIELD_REQUEST_SIZE |  # noqa
+                                 UCP_PARAM_FIELD_REQUEST_INIT)
+
+        # We always request UCP_FEATURE_WAKEUP even when in blocking mode
+        # See <https://github.com/rapidsai/ucx-py/pull/377>
+        ucp_params.features = (UCP_FEATURE_TAG |  # noqa
+                               UCP_FEATURE_WAKEUP |  # noqa
+                               UCP_FEATURE_STREAM)
+
+        ucp_params.request_size = sizeof(ucp_request)
+        ucp_params.request_init = (
+            <ucp_request_init_callback_t>ucp_request_reset
+        )
+
+        cdef ucp_config_t *config = _read_ucx_config(config_dict)
```

Comment:
Should we do some sort of `try...finally...` thing below to make sure this gets freed if some other error comes up?

Reply:
Fixed